sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/eventsourcedentity/ReflectiveEventSourcedEntityRouter.scala

```diff
@@ -17,31 +17,54 @@
 package kalix.springsdk.impl.eventsourcedentity
 
 import com.google.protobuf.any.{ Any => ScalaPbAny }
+import com.google.protobuf.{ Any => JavaPbAny }
+import kalix.javasdk.JsonSupport
+import kalix.javasdk.Metadata
 import kalix.javasdk.eventsourcedentity.{ CommandContext, EventSourcedEntity }
 import kalix.javasdk.impl.eventsourcedentity.EventSourcedEntityRouter
+import kalix.springsdk.impl.MethodInvoker
+import kalix.springsdk.impl.SpringSdkMessageCodec
 import kalix.springsdk.impl.{ CommandHandler, InvocationContext }
 
-import java.lang.reflect.Method
+import java.lang.reflect.ParameterizedType
 
 class ReflectiveEventSourcedEntityRouter[S, E <: EventSourcedEntity[S]](
     override protected val entity: E,
     commandHandlers: Map[String, CommandHandler],
-    eventHandlerMethods: Map[Class[_], Method])
+    eventHandlerMethods: Map[String, MethodInvoker],
+    messageCodec: SpringSdkMessageCodec)
     extends EventSourcedEntityRouter[S, E](entity) {
 
   private def commandHandlerLookup(commandName: String) =
-    commandHandlers.getOrElse(commandName, throw new RuntimeException(s"no matching method for '$commandName'"))
+    commandHandlers.getOrElse(
+      commandName,
+      throw new HandlerNotFoundException("command", commandName, commandHandlers.keySet))
 
-  private def eventHandlerLookup(eventClass: Class[_]) =
-    eventHandlerMethods.getOrElse(eventClass, throw new RuntimeException(s"no matching handler for '$eventClass'"))
+  private def eventHandlerLookup(eventName: String) =
+    eventHandlerMethods.getOrElse(
+      eventName,
+      throw new HandlerNotFoundException("event", eventName, commandHandlers.keySet))
 
   override def handleEvent(state: S, event: Any): S = {
 
-    entity._internalSetCurrentState(state)
+    _extractAndSetCurrentState(state)
 
-    eventHandlerLookup(event.getClass)
-      .invoke(entity, event.asInstanceOf[event.type])
-      .asInstanceOf[S]
+    event match {
+      case s: ScalaPbAny => // replaying event coming from proxy
+        // FIXME: where should we get the metadata from here?
```

Comment:
There is no metadata available here. We need to check if we have metadata where we call `handleEvent` and is positive, we need to see how it can be propagated further.

Reply:
would you agree this can be fixed on a follow-up PR? I can raise an issue for it.

Reply:
I see your ✅ now, I will raise a follow up issue for the above.